### ISSUE TITLE
Practical Guide to SQL Transaction Isolation

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4310]
Practical Guide to SQL Transaction Isolation<br>
<br>
http://bit.ly/2vrl22B<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<p>You may have seen isolation levels in the documentation for your database, felt mildly uneasy, and went on with life. Few day-to-day examples of using transactions really mention isolation. Most use the database&rsquo;s default isolation level and hope for the best. It&rsquo;s a fundamental topic to understand however and you&rsquo;ll feel more comfortable if you dedicate some time to study this guide.</p>
<p>I have assembled this information from academic papers, the PostgreSQL docs, and discussions with colleagues to answer not just <em>what</em> isolation levels are, but <em>when</em> to use them for maximum speed while preserving application correctness.</p>
<h2 id="basic-definitions">Basic Definitions</h2>
<p>To properly understand SQL isolation levels, we ought first to consider transactions themselves. The idea of a transaction comes from contract law: legal transactions must be atomic (either all provisions apply or none do), consistent (abiding by legal protocols), and durable (after commitment the parties cannot go back on their word). These properties are the A, C and D in the popular &ldquo;ACID&rdquo; acronym for database management systems. The final letter, &ldquo;I&rdquo; for isolation, is what this article is all about.</p>
<p>In databases, as opposed to law, a transaction is a group of operations that transform the database from one consistent state to another. This means that if all database consistency constraints were satisfied prior to running a transaction, then they will remain satisfied afterward.</p>
<p>Could the database have pushed this idea further and enforced constraints at each and every SQL data modification statement? Not with the available SQL commands. They are not expressive enough to allow the user to preserve consistency at every step. For instance, the classic task of transferring money from one bank account to another involves a temporarily inconsistent state after debiting one account but prior to crediting the other. For this reason transactions, and not statements, are treated as the units of consistency.</p>
<p>At this point we can imagine transactions running serially on the database, each waiting its turn for exclusive access to the data. In this orderly world the database would move from one consistent state to another, passing through brief periods of harmless inconsistency.</p>
<p>However the utopia of serialized transactions is infeasible for virtually any multi-user database systems. Imagine an airline database locking access for <em>everyone</em> while one customer books a flight.</p>
<p>Thankfully truly serialized transaction execution is usually unnecessary. Many transactions have nothing to do with one another because they update or read entirely separate information. The final result of running such transactions at the same time &ndash; of interleaving their commands &ndash; is indistinguishable from choosing to run one entire transaction before the other. In this case we call them <em>serializable.</em></p>
<p>However running transactions concurrently does pose the danger of conflicts. Without database oversight the transactions can interfere with each other&rsquo;s working data and can observe incorrect database state. This can cause incorrect query results and constraint violation.</p>
<p>Modern databases offer ways to automatically and selectively delay or retry commands in a transaction to prevent interference. The database offers several modes of increasing rigor for this prevention, called isolation levels. The &ldquo;higher&rdquo; levels employ more effective &ndash; but more costly &ndash; measures to detect or resolve conflicts.</p>
<p>Running concurrent transactions at different isolation levels allows application designers to balance concurrency and throughput. Lower isolation levels increase transaction concurrency at the risk of transactions observing certain kinds of incorrect database state.</p>
<p>Choosing the right level requires understanding which concurrent interactions pose a threat to the queries required by an application. As we will see, sometimes an application can get away with a lower than normal isolation level through manual actions like taking explicit locks.</p>
<p>Before examining isolation levels, let&rsquo;s take a stop at the zoo to see transaction problems in captivity. The literature calls these problems &ldquo;transaction phenomena.&rdquo;</p>
<h2 id="the-zoo-of-transaction-phenomena">The Zoo of Transaction Phenomena</h2>
<p>For each phenomenon we examine the telltale pattern of interleaved commands, see how it can be bad, and also note times when it can be tolerated or even used intentionally for desirable effects.</p>
<p>We&rsquo;ll use a shorthand notation for the actions of two transactions T1 and T2. Here are some examples:</p>
<ul>
<li>
<code>r1[x]</code> &ndash; T1 reads value/row x</li>
<li>
<code>w2[y]</code> &ndash; T2 writes value/row y</li>
<li>
<code>c1</code> &ndash; T1 commits</li>
<li>
<code>a2</code> &ndash; T2 aborts</li>
</ul>
<h3 id="dirty-writes">Dirty Writes</h3>
<p>Transaction T1 modifies an item, T2 further modifies it before T1 commits or rolls back.</p>
<div>
<img src="https://begriffs.com/images/isolation-dirty-write.jpg" alt="dirty write avatar">
<p>dirty write avatar</p>
</div>
<h4 id="pattern">Pattern</h4>
<div>
<img src="https://begriffs.com/images/isolation-diagram-dw.png" alt="w1[x]&hellip;w2[x]&hellip;(c1 or a1)">
<p>w1[x]&hellip;w2[x]&hellip;(c1 or a1)</p>
</div>
<h4 id="dangers">Dangers</h4>
<p>If dirty writes are permitted then it is not always possible for the database to roll back a transaction. Consider:</p>
<ul>
<li>{db in state A}</li>
<li>w1[x]</li>
<li>{db in state B}</li>
<li>w2[x]</li>
<li>{db in state C}</li>
<li>a1</li>
</ul>
<p>Should we go back to state A? No, because that would lose w2[x]. So we remain at state C. If c2 happens then we&rsquo;re good. However if a2 happens then what? We can&rsquo;t pick B or it would undo a1. But we can&rsquo;t pick C because that would undo a2. Reductio ad absurdum.</p>
<p>Because dirty writes break the atomicity of transactions, no relational database allows them at even the lowest isolation level. It&rsquo;s simply instructive to consider the problem in the abstract.</p>
<p>Dirty writes also allow a consistency violation. For instance suppose the constraint is x=y. The transactions T1 and T2 might individually preserve the constraint, but running together with a dirty write violate it:</p>
<ul>
<li>start, x = y = 0</li>
<li>w1[x=1] &hellip; w2[x=2] &hellip; w2[y=2] &hellip; w1[y=1]</li>
<li>now x = 2 &ne; 1 = y</li>
</ul>
<h4 id="legitimate-uses">Legitimate Uses</h4>
<p>There is no situation where dirty writes are useful, even as a shortcut. Hence no database allows them.</p>
<h3 id="dirty-reads">Dirty Reads</h3>
<p>A transaction reads data written by a concurrent uncommitted transaction. (As in the previous phenomenon, uncommitted data is called &ldquo;dirty.&rdquo;)</p>
<div>
<img src="https://begriffs.com/images/isolation-dirty-read.jpg" alt="dirty read avatar">
<p>dirty read avatar</p>
</div>
<h4 id="pattern-1">Pattern</h4>
<div>
<img src="https://begriffs.com/images/isolation-diagram-dr.png" alt="w1[x]&hellip;w2[x]&hellip;(c1 or a1)">
<p>w1[x]&hellip;w2[x]&hellip;(c1 or a1)</p>
</div>
<h4 id="dangers-1">Dangers</h4>
<p>Say T1 modifies a row, T2 reads it, then T1 rolls back. Now T2 is holding a row that &ldquo;never existed.&rdquo; Basing future decisions off of nonexistent data can be a bad idea.</p>
<p>Dirty reads also open the door for a constraint violation. Assume the constraint x=y. Also suppose T1 adds 100 to both x and y, and T2 doubles them both. Either transaction alone preserves x=y. However a dirty read of w1[x += 100], w2[x *= 2], w2[y *= 2], w1[y += 100] violates the constraint.</p>
<p>Finally, even if no concurrent transactions roll back, a transaction starting in the middle of another&rsquo;s operation can dirty read an inconsistent database state. We would prefer that transactions could count on being started in a consistent state.</p>
<h4 id="legitimate-uses-1">Legitimate Uses</h4>
<p>Dirty reads are useful when one transaction would like to spy on another, for instance during debugging or progress monitoring. For instance, repeatedly running COUNT(*) on a table from one transaction while another ingests data into it can show the ingestion speed/progress, but only if dirty reads are allowed.</p>
<p>Also this phenomenon won&rsquo;t happen during queries for historical information that has long ceased changing. No writes, no problems.</p>
<h3 id="non-repeatable-reads-and-read-skew">Non-Repeatable Reads, and Read Skew</h3>
<p>A transaction re-reads data it has previously read and finds that data has been modified by another transaction (that has committed since the initial read).</p>
<p>Note that this differs from a dirty read in that the other transaction has committed. Also this phenomenon requires two reads to manifest.</p>
<div>
<img src="https://begriffs.com/images/isolation-non-repeatable.jpg" alt="non-repeatable read avatar">
<p>non-repeatable read avatar</p>
</div>
<h4 id="pattern-2">Pattern</h4>
<div>
<img src="https://begriffs.com/images/isolation-diagram-nrr.png" alt="r1[x]&hellip;w2[x]&hellip;c2&hellip;r1[x]&hellip;c1">
<p>r1[x]&hellip;w2[x]&hellip;c2&hellip;r1[x]&hellip;c1</p>
</div>
<p>The form involving two values is called <em>read skew:</em></p>
<div>
<img src="https://begriffs.com/images/isolation-diagram-rs.png" alt="r1[x]&hellip;w2[x]&hellip;w2[y]&hellip;c2&hellip;r1[y]&hellip;(c1 or a1)">
<p>r1[x]&hellip;w2[x]&hellip;w2[y]&hellip;c2&hellip;r1[y]&hellip;(c1 or a1)</p>
</div>
<p>Non-repeatable read is a degenerate case where b=a.</p>
<h4 id="dangers-2">Dangers</h4>
<p>Like dirty reads, non-repeatable reads allow a transaction to read an inconsistent state. It happens in a slightly different way. Suppose the constraint is x=y.</p>
<ul>
<li>start, x = y = 0</li>
<li>r1[x] &hellip; w2[x=1] &hellip; w2[y=1] &hellip; c2 &hellip; r1[y]</li>
<li>From T1&rsquo;s perspective, x = 0 &ne; 1 = y</li>
</ul>
<p>T1 never read any dirty data, but T2 slipped in, changed values and committed between T1&rsquo;s reads. Notice this violation didn&rsquo;t even involve T1 re-reading the same value.</p>
<p>Read skew can cause constraint violations between two related elements. For instance, assume the constraint x+y &gt; 0. Then:</p>
<ul>
<li>start, x = y = 50</li>
<li>r1[x] &hellip; r1[y] &hellip; r2[x] &hellip; r2[y] &hellip; w1[y=-40] &hellip; w2[x=-40] &hellip; c1 &hellip; c2</li>
<li>T1 and T2 each observe x+y=10, but together they make it -80.</li>
</ul>
<p>Another constraint violation involving two values is that between a foreign key and its target. Read skew can mess that up too. For instance, T1 could read a row from table A pointing at table B. Then T2 can delete that row from B and commit. Now A believes the row exists in B but will be unable to read it.</p>
<p>This would be catastrophic when taking a database backup while other transactions are running since the observed state can be inconsistent and unsuitable for restoration.</p>
<h4 id="legitimate-uses-2">Legitimate Uses</h4>
<p>Doing non-repeatable reads allows access to the freshest committed data. This might be useful for large (or frequently repeated) aggregate reports when they can tolerate reading ephemeral constraint violations.</p>
<h3 id="phantom-reads">Phantom Reads</h3>
<p>A transaction re-executes a query returning a set of rows that satisfy a search condition and finds that the set of rows satisfying the condition has changed due to another recently-committed transaction.</p>
<p>This is similar to a non-repeatable read except it involves a changing collection matching a predicate rather than a single item.</p>
<div>
<img src="https://begriffs.com/images/isolation-phantom-read.jpg" alt="phantom read avatar">
<p>phantom read avatar</p>
</div>
<h4 id="pattern-3">Pattern</h4>
<div>
<img src="https://begriffs.com/images/isolation-diagram-pr.png" alt="r1[P]&hellip;w2[y in P]&hellip;c2&hellip;r1[P]">
<p>r1[P]&hellip;w2[y in P]&hellip;c2&hellip;r1[P]</p>
</div>
<h4 id="dangers-3">Dangers</h4>
<p>One situation is when a table contains rows that represent resource allocation (like employees and their salaries) where one transaction, &ldquo;the adjuster,&rdquo; increases the resources per row while another transaction inserts new rows. A phantom read will include the new rows, causing the adjuster to blow the budget.</p>
<p>Or a related example. Consider a constraint that says a set of job tasks determined by a predicate cannot exceed a sum of eight hours. T1 reads this predicate, determines the sum is only seven hours and adds a new task of one hour duration, while a concurrent transaction T2 does the same thing.</p>
<h4 id="legitimate-uses-3">Legitimate Uses</h4>
<p>Paginated search results may benefit from including brand new items as the page turns. Then again, inserted or deleted items can shift which items are on which pages as the user navigates.</p>
<h3 id="lost-update">Lost Update</h3>
<p>T1 reads an item. T2 updates it. T1 updates it possibly based on what it read, and commits. T2&rsquo;s update has been lost.</p>
<div>
<img src="https://begriffs.com/images/isolation-lost-update.jpg" alt="lost update avatar">
<p>lost update avatar</p>
</div>
<h4 id="pattern-4">Pattern</h4>
<div>
<img src="https://begriffs.com/images/isolation-diagram-lu.png" alt="r1[x]&hellip;w2[x]&hellip;w1[x]&hellip;c1">
<p>r1[x]&hellip;w2[x]&hellip;w1[x]&hellip;c1</p>
</div>
<h4 id="dangers-4">Dangers</h4>
<p>In some ways this almost doesn&rsquo;t feel like an anomaly. It can&rsquo;t violate database constraints because the end result is that some work simply hasn&rsquo;t been performed. Also it&rsquo;s similar to when an application blindly but serially updates the same value twice.</p>
<p>However it is an anomaly after all because there is zero chance for any other transaction to see the update, and T2&rsquo;s commit behaves like a rollback. In any serial execution <em>somebody</em> else would see the change, or at least could have if they checked.</p>
<p>Lost update has particularly bad effects when applications perform actions in the real world between the read and write.</p>
<p>For instance two people simultaneously try to buy the last available ticket for an event, spawning two transactions which read the there is one remaining ticket for sale. The app in separate threads queues emails with printable tickets, and updates the remaining ticket count to zero. After both updates happen, there are zero tickets remaining which is correct. However one of the customers has an email for a duplicate ticket.</p>
<p>Finally, note that the risk of lost updates increases when applications (usually through an ORM) update all columns in a row rather than just those columns that have changed since a read.</p>
<h4 id="legitimate-uses-4">Legitimate Uses</h4>
<p>Lost updates can&rsquo;t happen with atomic read-update statements like <code>UPDATE foo SET bar = bar + 1 WHERE id = 123;</code> because there is no way for another transaction to slip a write between the reading and incrementing of bar. The phenomenon happens when an application reads an item, performs internal calculations on it, and then writes a new value. But we&rsquo;ll get into that later.</p>
<p>Sometimes an application may be fine with missing some values in a history of updates. Perhaps sensors are rapidly overwriting a measurement from multiple threads and we want to read only a reasonably recent value. This case, albeit slightly contrived, can tolerate lost updates.</p>
<h3 id="write-skew">Write Skew</h3>
<p>Two concurrent transactions each determine what they are writing based on reading a data set which overlaps what the other is writing.</p>
<div>
<img src="https://begriffs.com/images/isolation-write-skew.jpg" alt="write skew avatar">
<p>write skew avatar</p>
</div>
<h4 id="pattern-5">Pattern</h4>
<div>
<img src="https://begriffs.com/images/isolation-diagram-ws.png" alt="r1[x]&hellip;r2[y]&hellip;w1[y]&hellip;w2[x]&hellip;(c1 and c2 occur)">
<p>r1[x]&hellip;r2[y]&hellip;w1[y]&hellip;w2[x]&hellip;(c1 and c2 occur)</p>
</div>
<p>Note that if b=a then we have a lost update.</p>
<h4 id="dangers-5">Dangers</h4>
<p>Write skew creates non-serializable transaction histories. Recall that this means no way of running the transactions one after another will produce the same result as their pathological interleaving.</p>
<p>The clearest example I&rsquo;ve seen is the case of the black and white rows. Copying verbatim from a PostgreSQL wiki: In this case there are rows with a color column containing &ldquo;black&rdquo; or &ldquo;white.&rdquo; Two users concurrently try to make all rows contain matching color values, but their attempts go in opposite directions. One is trying to update all white rows to black and the other is trying to update all black rows to white.</p>
<p>If these updates are run serially, all colors will match. However without any database protective measures the interleaved updates will simply reverse each other, leaving behind a mix of colors.</p>
<p>Write skew also breaks constraints. Suppose we constrain x + y &ge; 0. Then</p>
<ul>
<li>start, x = y = 100</li>
<li>r1[x] &hellip; r1[y] &hellip; r2[x] &hellip; r2[y] &hellip; w1[y=-y] &hellip; w2[x=-x]</li>
<li>now x+y = -200</li>
</ul>
<p>Both transactions read that x and y have value 100, so individually it&rsquo;s fine for each transaction to negate one the values, the total would still be non-negative. However negating both values results in x+y=-200, violating the constraint. For emotional gravity this is usually framed in terms of bank accounts where account balances are allowed to go negative as long as the sum of commonly held balances remains non-negative.</p>
<h3 id="read-only-serialization-anomaly">Read-Only Serialization Anomaly</h3>
<p>A transaction may see a control record updated to show that a batch has been completed but not see one of the detail records which is logically part of the batch because it read an earlier revision of the control record.</p>
<p>Whereas only two concurrent transaction sufficed to cause the previous anomalies, this one needs three. It attracted interest upon discovery in 2004 because it reveals a weakness in the snapshot isolation level (discussed later), as well as manifesting in the only transaction of the three that performs no writes.</p>
<div>
<img src="https://begriffs.com/images/isolation-read-only-anom.jpg" alt="read-only anomaly avatar">
<p>read-only anomaly avatar</p>
</div>
<h4 id="pattern-6">Pattern</h4>
<p>The transactions race to do three things,</p>
<ul>
<li>T1: generate report for current batch</li>
<li>T2: add new receipt to current batch</li>
<li>T3: make new batch the &ldquo;current&rdquo; one</li>
</ul>
<div>
<img src="https://begriffs.com/images/isolation-diagram-ro.png" alt="r2[b]&hellip;w3[b++]&hellip;r1[b]&hellip;r1[S_b]&hellip;w2[s in S_b]">
<p>r2[b]&hellip;w3[b++]&hellip;r1[b]&hellip;r1[S_b]&hellip;w2[s in S_b]</p>
</div>
<h4 id="dangers-6">Dangers</h4>
<p>The history demonstrated above is not serializable. Running the transactions serially has the invariant that after a report transaction has shown the total for a particular batch, subsequent transactions cannot change that total.</p>
<p>Database consistency remains intact with this anomaly but the results of the report are incorrect.</p>
<h4 id="legitimate-uses-5">Legitimate Uses</h4>
<p>Given that it took until 2004 for anyone to notice the phenomenon, it&rsquo;s not as likely to cause problems as prior phenomena in the zoo. There isn&rsquo;t any time that it&rsquo;s really desirable, but it may not be very severe either.</p>
<h3 id="others">Others?</h3>
<p>Have we identified all possible transaction phenomena? It can be tricky to tell; the ANSI SQL-92 standard thought they had covered everything with dirty read, non-repeatable read, and phantom read. It wasn&rsquo;t until 1995 that Berenson et al identified other serialization anomalies, and not until 2004 as noted that the read-only anomaly was documented.</p>
<p>The first relational databases used locking to manage concurrency. The SQL standard talked in terms of transaction phenomena rather than locks to allow non-lock-based implementations of the standard. However the reason that the standard writers failed to uncover other anomalies was because the three they identified were &ldquo;locks in disguise.&rdquo;</p>
<p>I personally don&rsquo;t know whether there are more undocumented transaction phenomena, but it seems doubtful. There are now numerous papers investigating the properties of serializability per se and it seems like the theoretical underpinnings are in place.</p>
<h2 id="isolation-levels">Isolation Levels</h2>
<p>Commercial databases provide concurrency control with a range of isolation levels which are in fact controlled serializability violations. An application picks lower levels in order to achieve higher performance. Higher performance means better transaction execution rate and shorter average transaction response time.</p>
<p>If you have understood the &ldquo;zoo&rdquo; of concurrency problems in the previous section, then you&rsquo;re well on your way to knowing enough to intelligently choose the right isolation level for your application. Without getting too deep into <em>how</em> the levels prevent different phenomena, here is <em>what</em> each prevents.</p>
<div>
<img src="https://begriffs.com/images/isolation-levels.png" alt="isolation levels graph">
<p>isolation levels graph</p>
</div>
<p>At the top, Serializable doesn&rsquo;t allow any phenomena. Following the arrows removes protection for the labeled anomaly.</p>
<p>The three nodes in blue are the levels actually offered by PostgreSQL. What&rsquo;s confusing is that the SQL spec recognizes a limited number of levels, so PostgreSQL maps the names from the spec to the actual levels supported:</p>
<table>
<thead>
<tr>
<th>What you ask for</th>
<th>What you get</th>
</tr>
</thead>
<tbody>
<tr>
<td>Serializable</td>
<td>Serializable</td>
</tr>
<tr>
<td>Repeatable Read</td>
<td>Snapshot Isolation</td>
</tr>
<tr>
<td>Read Committed</td>
<td>Read Committed</td>
</tr>
<tr>
<td>Read Uncommitted</td>
<td>Read Committed</td>
</tr>
</tbody>
</table>
<p>For example:</p>
<div>
<pre>
<code><span>BEGIN</span> <span>ISOLATION</span> <span>LEVEL</span> REPEATABLE <span>READ</span>; <span>-- Now we're in snapshot isolation</span></code>
</pre>
</div>
<p>Read committed is the default level, so imagine the concurrency problems your existing apps might be experiencing if you haven&rsquo;t taken precautions.</p>
<h3 id="optimistic-vs-pessimistic">Optimistic vs Pessimistic</h3>
<p>As mentioned, we&rsquo;re not going to dive into how each of PostgreSQL&rsquo;s isolation levels prevents concurrency phenomena, but we do need to understand there are two general approaches: optimistic and pessimistic concurrency control. It&rsquo;s important because each approach requires different application programming techniques.</p>
<p>Pessimistic concurrency control takes locks on database rows to force transactions to wait their turn reading and writing. It is &ldquo;pessimistic&rdquo; because it always takes the time to acquire and release locks, assuming gloomily that there will be contention.</p>
<p>Optimistic control doesn&rsquo;t bother taking locks, it just puts each transaction into a separate snapshot of the database state and watches for any contention to happen. If one transaction interferes with another, the database halts the offender and erases its work. This ends up being efficient when interference is rare.</p>
<p>The number of conflicts encountered depends on several factors:</p>
<ul>
<li>Contention for individual rows. The potential for conflicts increases as the number of transactions attempting to update the same row increases.</li>
<li>Number of rows read in isolation levels which prevent non-repeatable reads. The more rows read, the greater the chance that some of these rows are updated by concurrent transactions.</li>
<li>Size of the scan ranges used in isolation levels which prevent phantom reads. The larger the scan ranges, the higher the chance that concurrent transactions will introduce phantom rows.</li>
</ul>
<p>In PostgreSQL, two levels use optimistic concurrency control: the repeatable read (which is actually snapshot isolation) and serializable levels. <strong>These levels are not magic fairy dust that you sprinkle on an unsafe app to fix its problems.</strong> They require modifying application logic.</p>
<p>An application interacting with PostgreSQL using an isolation level with optimistic concurrency control must be built carefully. Remember that nothing is certain until commit, all work can be erased at a moment&rsquo;s notice. The app must be prepared to detect when its query has been halted with error 40001 (aka <code>serialization_failure</code>), and retry the transaction. Also an application should never perform an irreversible real-world action during such a transaction. The app must guard that kind of action with a pessimistic lock, or perform the action at the conclusion of a successful commit.</p>
<p>Also one might imagine catching the serialization exceptions and retrying them within a PL/pgSQL function, but the retry cannot happen there. The whole function runs <em>within</em> a transaction, and loses control of execution before commit is called. Unfortunately it is mostly at commit time when serialization errors happens, too late for the function to catch them.</p>
<p>Retries must be made by the database client. Many languages provide helper libraries for the task. Here are a few.</p>
<p>Because re-doing transactions can be wasteful, it&rsquo;s good to remember that simple transactions with limited duration are most effective at avoiding lost work.</p>
<h3 id="compensating-for-low-isolation-levels">Compensating for Low Isolation Levels</h3>
<p>Generally it&rsquo;s best to use an isolation level which protects against any anomalies that would disturb your queries. Let the database do what it does best. However if you believe that only certain anomalies will be happen in your situation, you can choose to use a lower isolation level with pessimistic locking.</p>
<p>For example we can prevent lost updates in a read committed transaction by taking a lock on the row between reading and updating it. Just add &ldquo;FOR UPDATE&rdquo; to the select statement.</p>
<div>
<pre>
<code><span>BEGIN</span>; <span>SELECT</span> * <span>FROM</span> player <span>WHERE</span> <span>id</span> = <span>42</span> <span>FOR</span> <span>UPDATE</span>; <span>-- some kind of game logic here</span> <span>UPDATE</span> player <span>SET</span> score = <span>853</span> <span>WHERE</span> <span>id</span> = <span>42</span>; <span>COMMIT</span>;</code>
</pre>
</div>
<p>Any other transaction which tries to also select that row for update will block until the first transaction is complete. This select for update trick is even useful in serializable transactions to avoid serialization errors which would require a retry, especially if you intend to perform non-idempotent application actions.</p>
<p>Finally, you can take a calculated risk in going for a lower level. The main reason for snapshot isolation&rsquo;s adoption is that it allows better performance than serializability, yet still avoids most of the concurrency anomalies that serializability avoids. If write skew is not anticipated for your situation then you can turn the level down to snapshot.</p>
<h2 id="sources-and-further-reading">Sources and Further Reading</h2>
<p>Thanks to a number of people who have advised me in the writing of this article.</p>
<ul>
<li>On the #postgresql Freenode IRC channel: Andrew Gierth (RhodiumToad) and Vik Fearing (xocolatl)</li>
<li>Personal conversations: Marco Slot, Andres Freund, Samay Sharma, and Daniel Farina from <a href="https://www.citusdata.com/">Citus Data</a>
</li>
</ul>
<p>Further reading</p>
<h3 id="related-posts">Related Posts</h3>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via begriffs.com https://begriffs.com<br>
<br>
June 11, 2019 at 02:34PM
